### PR TITLE
[Merged by Bors] - feat: port `GroupTheory/Submonoid/Basic`

### PR DESCRIPTION
**Pull request by ADedecker**

### CODE REVIEW

File: Mathlib/GroupTheory/Submonoid/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+    (m ∈ ⨆ i, p i) ↔ ∀ N, (∀ i, p i ≤ N) → m ∈ N := by
+  rw [← closure_singleton_le_iff_mem, le_supᵢ_iff]
+  simp only [closure_singleton_le_iff_mem]
+#align submonoid.mem_supr Submonoid.mem_supr
+
+@[to_additive]
+theorem supr_eq_closure {ι : Sort _} (p : ι → Submonoid M) :
+    (⨆ i, p i) = Submonoid.closure (⋃ i, (p i : Set M)) := by
+  simp_rw [Submonoid.closure_unionᵢ, Submonoid.closure_eq]
+#align submonoid.supr_eq_closure Submonoid.supr_eq_closure
```
Comment (ChrisHughes24) [inline Mathlib/GroupTheory/Submonoid/Basic.lean:526]:
```suggestion
@[to_additive]
theorem mem_supᵢ {ι : Sort _} (p : ι → Submonoid M) {m : M} :
    (m ∈ ⨆ i, p i) ↔ ∀ N, (∀ i, p i ≤ N) → m ∈ N := by
  rw [← closure_singleton_le_iff_mem, le_supᵢ_iff]
  simp only [closure_singleton_le_iff_mem]
#align submonoid.mem_supr Submonoid.mem_supᵢ

@[to_additive]
theorem supᵢ_eq_closure {ι : Sort _} (p : ι → Submonoid M) :
    (⨆ i, p i) = Submonoid.closure (⋃ i, (p i : Set M)) := by
  simp_rw [Submonoid.closure_unionᵢ, Submonoid.closure_eq]
#align submonoid.supr_eq_closure Submonoid.supᵢ_eq_closure
```

**Review by ChrisHughes24 — APPROVED** [review]

**Comment by ChrisHughes24** [comment]
bors d+

**Comment by urkud** [comment]
* The loop in `pow_mem` can be resolved if you add `: Prop` to `MulMemClass` etc.
* `eq_mlocus` etc were there for submonoids so that we can use non-`m` versions for subgroups.

**Comment by urkud** [comment]
bors d-

**Comment by ADedecker** [comment]
> The loop in pow_mem can be resolved if you add : Prop to MulMemClass etc.

That didn't make the situation better on my side :thinking: 
What did however was changing some arguments from `[]` to `{}` in `SubmonoidClass.toOneMemClass`

> `eq_mlocus` etc were there for submonoids so that we can use non-`m` versions for subgroups.

Yes I know that, but I saw you removed some of them when porting `GroupTheory.Subsemigroup.Basic`. Is it because they weren't supposed to be there in the first place? Anyway, I'll add them back

**Comment by urkud** [comment]
> > `eq_mlocus` etc were there for submonoids so that we can use non-`m` versions for subgroups.
> 
> Yes I know that, but I saw you removed some of them when porting `GroupTheory.Subsemigroup.Basic`. Is it because they weren't supposed to be there in the first place? Anyway, I'll add them back

I think that somebody copied them from `submonoid` to `subsemigroup` in `mathlib`. IMHO, they shouldn't be there.

**Comment by urkud** [comment]
Thanks! I think that these typeclasses should be moved to `Prop` but this can wait for leanprover-community/mathlib#18015 to be merged.
bors d+

**Comment by semorrison** [comment]
@ADedecker, @urkud, @ChrisHughes24, just a heads up that this PR has lots of missing #aligns for the to_additive versions of statements. We should be checking for this when reviewing.

**Pull request closed without merging**